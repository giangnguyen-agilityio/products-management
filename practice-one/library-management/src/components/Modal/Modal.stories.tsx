import { type StoryFn, type Meta } from "@storybook/react";
import Modal from "./index";

export default {
  title: "Example/Modal",
  component: Modal,
  argTypes: {
    closeModal: { action: "Close modal button clicked" }
  },
} as Meta<typeof Modal>;

const Template: StoryFn<typeof Modal> = (args) => (
  <Modal {...args} />
);

export const Default = Template.bind({});
Default.args = {
    showModal: true,
    modalType: 'add'
};
