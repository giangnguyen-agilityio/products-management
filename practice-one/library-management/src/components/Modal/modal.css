.modal-overlay {
  width: 100%;
  height: 100%;
  background-color: var(--bg-overlay-color);
  position: absolute;
  padding: 10% 0;
  top: 0;
  right: 0;
}

.modal-container {
  width: 600px;
  display: flex;
  flex-direction: column;
  margin: 0 auto;
  border: 2px solid var(--primary-color);
  border-radius: 20px;
  background-color: var(--secondary-text-color);
  overflow: hidden;
  padding: 20px;
}

.modal-delete {
  width: 500px;
}

.modal-header {
  display: flex;
  justify-content: space-between;
  align-items: center;
  text-transform: uppercase;
}

.modal-title {
  font-family: var(--tertiary-font);
  color: var(--primary-color);
  font-size: 22px;
}

.close-modal-btn {
  padding: 8px;
  border: 3px solid var(--secondary-color);
  background-color: unset;
  color: var(--secondary-color);
  font-family: var(--secondary-font);
  font-size: 18px;
  transition: all 0.3s ease-in-out;
  border-radius: 50%;
}

.close-modal-btn:hover {
  color: var(--primary-text-color);
  border: 3px solid var(--primary-text-color);
}

/* Media Query for Laptop Devices */
@media only screen and (max-width: 1024px) {
  .modal-container {
    width: 100%;
  }
}

/* Media Query for Mobile Devices */
@media only screen and (max-width: 480px) {
  .modal-overlay {
    padding: 0;
  }
}
