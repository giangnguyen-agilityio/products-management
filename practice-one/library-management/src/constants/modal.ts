export enum MODAL {
  ADD = 'add',
  EDIT = 'edit',
  DELETE = 'delete',
}

export const MODAL_TITLE = {
  MODAl_ADD_BOOK_TITLE: 'Add a new book',
  MODAL_EDIT_BOOK_TITLE: 'Edit the book',

  MODAL_EDIT_HIRE_REQUEST: 'Edit the hire request',
}
