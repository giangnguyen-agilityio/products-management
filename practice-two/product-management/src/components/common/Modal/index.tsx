// Libraries
import React, { memo } from 'react'
import {
  Modal as ChakraModal,
  ModalCloseButton,
  ModalOverlay,
  ModalContent,
  ModalHeader,
  ModalBody,
} from '@chakra-ui/react'

// Interface for the props that the Modal component receives
export interface ModalProps {
  isOpen: boolean
  onCloseModal: () => void
  children?: React.ReactNode
}

// The Modal component
const Modal: React.FC<ModalProps> = ({ isOpen, onCloseModal, children }) => {
  return (
    <ChakraModal isOpen={isOpen} onClose={onCloseModal} size="lg">
      <ModalOverlay />
      <ModalContent background="background">
        <ModalHeader>
          {/* Close button */}
          <ModalCloseButton
            borderRadius="full"
            border="2px solid"
            borderColor="transparent"
            color="primary"
            _hover={{
              borderColor: 'primary',
            }}
          />
        </ModalHeader>

        <ModalBody>{children}</ModalBody>
      </ModalContent>
    </ChakraModal>
  )
}

export default memo(Modal)
