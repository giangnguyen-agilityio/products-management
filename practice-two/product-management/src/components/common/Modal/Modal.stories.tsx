import type { Meta, StoryObj } from '@storybook/react'
import Modal from './index'
import { MODAL } from '@constants'

const meta = {
  title: 'Practice Two/Modal Component',
  component: Modal,
  argTypes: {
    onCloseModal: { action: 'Close the modal button clicked' },
    onAdd: { action: 'Add a new product button clicked' },
    onEdit: { action: 'Edit a product button clicked' },
  },
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
} satisfies Meta<typeof Modal>

export default meta
type Story = StoryObj<typeof meta>

const productDataSample = {
  id: 'P01',
  name: 'iPhone 14 Pro Max',
  image: 'https://assets.nicepagecdn.com/d2cc3eaa/3159880/images/yttyy.jpg',
  discount: 11,
  oldPrice: 1263.8,
  newPrice: 1116.31,
  rate: 4.5,
  description:
    "The iPhone 14 Pro Max is Apple's flagship smartphone, featuring a Super Retina XDR display, A16 Bionic chip for powerful performance, and an advanced triple-camera system for exceptional photography. With premium design, Face ID, and MagSafe technology, it offers a top-tier user experience.",
}

export const ModalAdd: Story = {
  args: {
    isOpen: false,
    modalType: MODAL.ADD,
  },
}

export const ModalEdit: Story = {
  args: {
    isOpen: false,
    modalType: MODAL.EDIT,
    productData: productDataSample,
  },
}
