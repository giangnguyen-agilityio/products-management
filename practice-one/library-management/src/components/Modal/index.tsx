import React, {memo, useCallback, useEffect} from 'react'
import Typography from '@components/commons/Typography'
import Button from '@components/commons/Button'
import {IoClose} from 'react-icons/io5'
import {MODAL} from '@constants/modal'
import './modal.css'

interface ModalProps {
  closeModal: () => void
  showModal: boolean
  modalType: MODAL.ADD | MODAL.EDIT | MODAL.DELETE
  modalTitle: React.ReactNode
  children: React.ReactNode
}

const Modal: React.FC<ModalProps> = props => {
  const {modalType, showModal, modalTitle, children, closeModal} = props

  // Function to handle close the modal
  const handleCloseModal = useCallback((): void => {
    closeModal()
    document.removeEventListener('keydown', handleKeyDown)
  }, [closeModal])

  // Function to handle keydown events
  const handleKeyDown = useCallback(
    (event: KeyboardEvent): void => {
      if (event.key === 'Escape') {
        handleCloseModal()
      }
    },
    [handleCloseModal]
  )

  // Function to handle button close modal clicked
  const handleCloseModalButtonClick = useCallback((): void => {
    handleCloseModal()
  }, [handleCloseModal])

  useEffect(() => {
    if (showModal) {
      document.addEventListener('keydown', handleKeyDown)
    }
    return () => {
      document.removeEventListener('keydown', handleKeyDown)
    }
  }, [showModal, handleKeyDown])

  return showModal ? (
    <>
      <div className="modal-overlay">
        <div
          className={`modal-container ${
            modalType !== MODAL.ADD && modalType !== MODAL.EDIT
              ? 'modal-delete'
              : ''
          }`}
        >
          <div className="modal-header">
            <Typography variant="h2" className="modal-title">
              {modalTitle}
            </Typography>
            <Button
              size="large"
              variant="primary"
              className="close-modal-btn"
              aria-label="Close the modal button"
              onClick={handleCloseModalButtonClick}
            >
              <IoClose size={20} />
            </Button>
          </div>
          <div className="modal-content">{children}</div>
        </div>
      </div>
    </>
  ) : null
}

export default memo(Modal)
