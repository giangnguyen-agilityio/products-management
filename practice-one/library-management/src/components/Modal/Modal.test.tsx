import React from 'react'
import {render, fireEvent} from '@testing-library/react'
import {MODAL} from '@constants/modal'
import Modal from '.'

describe('Modal component', () => {
  it('should render the modal when `showModal` is true', () => {
    const {container} = render(
      <Modal
        closeModal={() => {}}
        showModal={true}
        modalType={MODAL.ADD}
        modalTitle="Test Modal"
      >
        Modal Content
      </Modal>
    )

    const modalContainer = container.querySelector('.modal-container')
    expect(modalContainer).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it('should not render the modal when `showModal` is false', () => {
    const {container} = render(
      <Modal
        closeModal={() => {}}
        showModal={false}
        modalType={MODAL.ADD}
        modalTitle="Test Modal"
      >
        Modal Content
      </Modal>
    )

    const modalContainer = container.querySelector('.modal-container')
    expect(modalContainer).not.toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it('should render the correct modal type based on `modalType` prop', () => {
    const {container} = render(
      <Modal
        closeModal={() => {}}
        showModal={true}
        modalType={MODAL.DELETE}
        modalTitle="Test Modal"
      >
        Modal Content
      </Modal>
    )

    const modalContainer = container.querySelector('.modal-delete')
    expect(modalContainer).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it('should call `closeModal` when the close button is clicked', () => {
    const closeModalMock = jest.fn()
    const {container} = render(
      <Modal
        closeModal={closeModalMock}
        showModal={true}
        modalType={MODAL.ADD}
        modalTitle="Test Modal"
      >
        Modal Content
      </Modal>
    )

    const closeButton = container.querySelector('.close-modal-btn')
    fireEvent.click(closeButton!)
    expect(closeModalMock).toHaveBeenCalledTimes(1)
  })

  it('should call `closeModal` when the Escape key is pressed', () => {
    const closeModalMock = jest.fn()
    const {container} = render(
      <Modal
        closeModal={closeModalMock}
        showModal={true}
        modalType={MODAL.ADD}
        modalTitle="Test Modal"
      >
        Modal Content
      </Modal>
    )

    const modalOverlay = container.querySelector('.modal-overlay')
    fireEvent.keyDown(modalOverlay!, {key: 'Escape'})
    expect(closeModalMock).toHaveBeenCalledTimes(1)
  })

  it('should not call `closeModal` when a key other than Escape is pressed', () => {
    const closeModalMock = jest.fn()
    const {container} = render(
      <Modal
        closeModal={closeModalMock}
        showModal={true}
        modalType={MODAL.ADD}
        modalTitle="Test Modal"
      >
        Modal Content
      </Modal>
    )

    const modalOverlay = container.querySelector('.modal-overlay')
    fireEvent.keyDown(modalOverlay!, {key: 'Enter'})
    expect(closeModalMock).not.toHaveBeenCalled()
  })
})
