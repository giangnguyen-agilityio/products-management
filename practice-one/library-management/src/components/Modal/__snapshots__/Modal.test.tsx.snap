// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component should not render the modal when \`showModal\` is false 1`] = `<div />`;

exports[`Modal component should render the correct modal type based on \`modalType\` prop 1`] = `
<div>
  <div
    class="modal-overlay"
  >
    <div
      class="modal-container modal-delete"
    >
      <div
        class="modal-header"
      >
        <h2
          class="text-h2 text-default modal-title"
        >
          Test Modal
        </h2>
        <button
          class="btn btn-primary btn-large close-modal-btn"
          type="button"
        >
          <svg
            fill="currentColor"
            height="20"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M289.94 256l95-95A24 24 0 00351 127l-95 95-95-95a24 24 0 00-34 34l95 95-95 95a24 24 0 1034 34l95-95 95 95a24 24 0 0034-34z"
            />
          </svg>
        </button>
      </div>
      <div
        class="modal-content"
      >
        Modal Content
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal component should render the modal when \`showModal\` is true 1`] = `
<div>
  <div
    class="modal-overlay"
  >
    <div
      class="modal-container "
    >
      <div
        class="modal-header"
      >
        <h2
          class="text-h2 text-default modal-title"
        >
          Test Modal
        </h2>
        <button
          class="btn btn-primary btn-large close-modal-btn"
          type="button"
        >
          <svg
            fill="currentColor"
            height="20"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M289.94 256l95-95A24 24 0 00351 127l-95 95-95-95a24 24 0 00-34 34l95 95-95 95a24 24 0 1034 34l95-95 95 95a24 24 0 0034-34z"
            />
          </svg>
        </button>
      </div>
      <div
        class="modal-content"
      >
        Modal Content
      </div>
    </div>
  </div>
</div>
`;
